File: packages/deon-java/source/deon/Network.java
package deon;

import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;

/**
 * Reading a resource over HTTP once the network has been granted (specification 9). The stdlib client
 * carries this — no third-party dependency, and TLS comes with it. An import asks for Deon or JSON; an
 * injection asks for anything; a link asks for Deon and nothing else. A non-2xx status is
 * DEON_RESOURCE_IO — it was allowed and it failed, which is a different thing from never having been
 * allowed. An empty token sends no header, because {@code Bearer } is a credential-shaped nothing.
 */
final class Network {
    private static final HttpClient CLIENT = HttpClient.newBuilder()
            .followRedirects(HttpClient.Redirect.NORMAL)
            .build();

    static String httpGet(String target, String kind, String token, Span span) {
        String accept = switch (kind) {
            case "import" -> "text/plain,application/json,application/deon";
            case "link" -> "application/deon";
            default -> "*/*";
        };

        HttpRequest.Builder builder;
        try {
            builder = HttpRequest.newBuilder(URI.create(target)).GET().header("Accept", accept);
        } catch (IllegalArgumentException e) {
            throw new DeonException(Code.RESOURCE_IO, "Unable to reach resource '" + target + "'.", span);
        }
        if (token != null && !token.isEmpty()) {
            builder.header("Authorization", "Bearer " + token);
        }

        HttpResponse<byte[]> response;
        try {
            response = CLIENT.send(builder.build(), HttpResponse.BodyHandlers.ofByteArray());
        } catch (Exception e) {
            throw new DeonException(Code.RESOURCE_IO, "Unable to reach resource '" + target + "'.", span);
        }

        if (response.statusCode() < 200 || response.statusCode() >= 300) {
            throw new DeonException(Code.RESOURCE_IO, "Resource '" + target + "' returned a non-success status.", span);
        }
        byte[] body = response.body();
        if (!Interpreter.isValidUtf8(body)) {
            throw new DeonException(Code.RESOURCE_FORMAT, "The resource '" + target + "' is not valid UTF-8.", span);
        }
        return new String(body, StandardCharsets.UTF_8);
    }
}

File: packages/deon-go/network.go
package deon

import (
	"io"
	"net/http"
)

// httpGet reads a resource over HTTP once the network has been granted (specification 9). An import
// asks for Deon or JSON; an injection asks for anything. A non-2xx status is DEON_RESOURCE_IO — it was
// allowed and it failed, which is a different thing from never having been allowed. An empty token
// sends no header, because `Bearer ` is a credential-shaped nothing.
func httpGet(target, kind, token string, span Span) string {
	request, err := http.NewRequest(http.MethodGet, target, nil)
	if err != nil {
		fail(ResourceIO, "Unable to reach resource '"+target+"'.", span)
	}

	if kind == "import" {
		request.Header.Set("Accept", "text/plain,application/json,application/deon")
	} else {
		request.Header.Set("Accept", "*/*")
	}
	if token != "" {
		request.Header.Set("Authorization", "Bearer "+token)
	}

	response, err := http.DefaultClient.Do(request)
	if err != nil {
		fail(ResourceIO, "Unable to reach resource '"+target+"'.", span)
	}
	defer response.Body.Close()

	if response.StatusCode < 200 || response.StatusCode >= 300 {
		fail(ResourceIO, "Resource '"+target+"' returned a non-success status.", span)
	}

	body, err := io.ReadAll(response.Body)
	if err != nil {
		fail(ResourceIO, "Unable to read resource '"+target+"'.", span)
	}
	if !isValidUTF8(string(body)) {
		fail(ResourceFormat, "The resource '"+target+"' is not valid UTF-8.", span)
	}
	return string(body)
}
